### PR TITLE
UCT/GTEST: tcp_sockcm - implement the disconnect flow

### PR DESCRIPTION
**Pull request by alinask**

### CODE REVIEW

File: src/uct/tcp/tcp_sockcm_ep.c [hunk 1]
```diff
@@ -365,6 +472,12 @@ ucs_status_t uct_tcp_sockcm_ep_server_handle_data_received(uct_tcp_sockcm_ep_t *
                                                   UCT_TCP_SOCKCM_EP_DATA_RECEIVED));
 
         ucs_assert(hdr->length == 0);
+
+        if (cep->state & UCT_TCP_SOCKCM_EP_GOT_DISCONNECT) {
```
Comment (yosefe) [inline src/uct/tcp/tcp_sockcm_ep.c:476]: aren't we checking this already in uct_tcp_sockcm_ep_recv() ?

File: src/uct/tcp/tcp_sockcm_ep.c [hunk 2]
```diff
@@ -23,9 +23,114 @@ uct_tcp_sockcm_t *uct_tcp_sockcm_ep_get_cm(uct_tcp_sockcm_ep_t *cep)
...
+    return cep->state & (UCT_TCP_SOCKCM_EP_DISCONNECTING |
+                         UCT_TCP_SOCKCM_EP_GOT_DISCONNECT);
+}
+
+static void uct_tcp_sockcm_ep_client_connect_cb(uct_tcp_sockcm_ep_t *cep,
+                                                uct_cm_remote_data_t *remote_data,
+                                                ucs_status_t status)
+{
+    cep->state |= UCT_TCP_SOCKCM_EP_CLIENT_CONNECTED_CB_INVOKED;
+    uct_cm_ep_client_connect_cb(&cep->super, remote_data, status);
```
[the remark this answers is not in the document]
Comment (alinask) [reply src/uct/tcp/tcp_sockcm_ep.c:65]: if we decide to unite the flags (move then to cm_base) then yes we can unite.

File: src/uct/tcp/tcp_sockcm_ep.c [hunk 3]
```diff
@@ -23,9 +23,114 @@ uct_tcp_sockcm_t *uct_tcp_sockcm_ep_get_cm(uct_tcp_sockcm_ep_t *cep)
...
+    int ret;
+
+    ucs_debug("ep %p (fd=%d state=%d) disconnecting from peer :%s", cep, cep->fd,
+              cep->state, uct_tcp_sockcm_cm_ep_peer_addr_str(cep, peer_str,
+                                                             UCS_SOCKADDR_STRING_LEN));
+
+    UCS_ASYNC_BLOCK(tcp_sockcm->super.iface.worker->async);
+
+    if (ucs_unlikely(cep->state & UCT_TCP_SOCKCM_EP_DISCONNECTING)) {
+        if (cep->state & UCT_TCP_SOCKCM_EP_GOT_DISCONNECT) {
```
[the remark this answers is not in the document]
Comment (alinask) [reply src/uct/tcp/tcp_sockcm_ep.c:90]: but we should return different statuses for different cases of a duplicate call to uct_ep_disconnect (UCS_ERR_NOT_CONNECTED or UCS_INPROGRESS)

File: src/uct/tcp/tcp_sockcm_ep.c [hunk 4]
```diff
@@ -333,6 +433,13 @@ static ucs_status_t uct_tcp_sockcm_ep_client_invoke_connect_cb(uct_tcp_sockcm_ep
     uct_cm_remote_data_t remote_data;
     ucs_status_t status;
 
+    if (cep->state & UCT_TCP_SOCKCM_EP_GOT_DISCONNECT) {
```
[the remark this answers is not in the document]
Comment (alinask) [reply src/uct/tcp/tcp_sockcm_ep.c:436]: what if we got here from an old read event from the missed queue?

File: src/uct/tcp/tcp_sockcm_ep.c (hunk 1 above)
Comment (alinask) [reply src/uct/tcp/tcp_sockcm_ep.c:476]: right. so actually it also applies to your previous comment - https://github.com/openucx/ucx/pull/5265#discussion_r437322650

File: src/uct/tcp/tcp_sockcm_ep.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/tcp/tcp_sockcm_ep.c:90]: ok

File: src/uct/tcp/tcp_sockcm_ep.c (hunk 1 above)
Comment (yosefe) [reply src/uct/tcp/tcp_sockcm_ep.c:476]: yes, it should be enough to check once

File: src/uct/tcp/tcp_sockcm_ep.c [hunk 5]
```diff
@@ -472,12 +465,6 @@ ucs_status_t uct_tcp_sockcm_ep_server_handle_data_received(uct_tcp_sockcm_ep_t *
                                                   UCT_TCP_SOCKCM_EP_DATA_RECEIVED));
 
         ucs_assert(hdr->length == 0);
-
-        if (cep->state & UCT_TCP_SOCKCM_EP_GOT_DISCONNECT) {
```
Comment (yosefe) [inline src/uct/tcp/tcp_sockcm_ep.c]: replace it by assert?

File: src/uct/tcp/tcp_sockcm_ep.c [hunk 6]
```diff
@@ -433,13 +433,6 @@ static ucs_status_t uct_tcp_sockcm_ep_client_invoke_connect_cb(uct_tcp_sockcm_ep
     uct_cm_remote_data_t remote_data;
     ucs_status_t status;
 
-    if (cep->state & UCT_TCP_SOCKCM_EP_GOT_DISCONNECT) {
```
Comment (yosefe) [inline src/uct/tcp/tcp_sockcm_ep.c]: replace it by assert?

File: src/uct/tcp/tcp_sockcm_ep.c [hunk 7]
```diff
@@ -23,40 +23,163 @@ uct_tcp_sockcm_t *uct_tcp_sockcm_ep_get_cm(uct_tcp_sockcm_ep_t *cep)
...
+                                                      char *buf, size_t max)
+{
+    struct sockaddr_storage remote_dev_addr = {0}; /* Suppress Clang false-positive */
+    socklen_t remote_dev_addr_len;
+    ucs_status_t status;
+
+    /* get the device address of the remote peer associated with the connected fd */
+    status = ucs_socket_getpeername(cep->fd, &remote_dev_addr, &remote_dev_addr_len);
+    if (status != UCS_OK) {
+        return NULL;
```
Comment (evgeny-leksikov) [inline src/uct/tcp/tcp_sockcm_ep.c:36]: is it valid to pass NULL pointer to "%s" format print? maybe print ucs_status_string(status) to buf?

File: src/uct/tcp/tcp_sockcm_ep.c [hunk 8]
```diff
@@ -23,40 +23,163 @@ uct_tcp_sockcm_t *uct_tcp_sockcm_ep_get_cm(uct_tcp_sockcm_ep_t *cep)
...
+    ucs_close_fd(fd);
+}
+
+static int uct_tcp_sockcm_ep_is_connected(uct_tcp_sockcm_ep_t *cep)
+{
+    return cep->state & (UCT_TCP_SOCKCM_EP_CLIENT_CONNECTED_CB_INVOKED |
+                         UCT_TCP_SOCKCM_EP_SERVER_NOTIFY_CB_INVOKED);
+}
+
+static int uct_tcp_sockcm_ep_is_in_disconnect(uct_tcp_sockcm_ep_t *cep)
```
Comment (evgeny-leksikov) [inline src/uct/tcp/tcp_sockcm_ep.c:54]: uct_tcp_sockcm_ep_is_disconnecting?

File: src/uct/tcp/tcp_sockcm_ep.c [hunk 9]
```diff
@@ -23,40 +23,163 @@ uct_tcp_sockcm_t *uct_tcp_sockcm_ep_get_cm(uct_tcp_sockcm_ep_t *cep)
...
 ucs_status_t uct_tcp_sockcm_ep_disconnect(uct_ep_h ep, unsigned flags)
 {
-    return UCS_ERR_NOT_IMPLEMENTED;
+    uct_tcp_sockcm_ep_t *cep     = ucs_derived_of(ep, uct_tcp_sockcm_ep_t);
+    uct_tcp_sockcm_t *tcp_sockcm = uct_tcp_sockcm_ep_get_cm(cep);
+    char peer_str[UCS_SOCKADDR_STRING_LEN];
+    ucs_status_t status;
+    int ret;
+
+    ucs_debug("ep %p (fd=%d state=%d) disconnecting from peer :%s", cep, cep->fd,
```
Comment (evgeny-leksikov) [inline src/uct/tcp/tcp_sockcm_ep.c:83]: pls move under lock since IIUC `cep->state` can be changed from other thread

**Comment by alinask** [comment]
@yosefe Is this good to go?

**Pull request merged**